packages/vite/src/node/plugins/assetImportMetaUrl.ts

```diff
@@ -120,11 +93,38 @@ export function assetImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
             )
             builtUrl = url
           }
-          s.update(
-            index,
-            index + exp.length,
-            `new URL(${JSON.stringify(builtUrl)}, self.location)`,
-          )
+
+          // potential dynamic template string
+          if (rawUrl[0] === '`' && /\$\{/.test(rawUrl)) {
+            const ast = this.parse(rawUrl)
+            const templateLiteral = (ast as any).body[0].expression
+            if (templateLiteral.expressions.length) {
+              const pattern = buildGlobPattern(templateLiteral)
+              if (pattern.startsWith('**')) {
+                // don't transform for patterns like this
+                // because users won't intend to do that in most cases
+                continue
+              }
+
+              // Note: native import.meta.url is not supported in the baseline
+              // target so we use the global location here. It can be
+              // window.location or self.location in case it is used in a Web Worker.
+              // @see https://developer.mozilla.org/en-US/docs/Web/API/Window/self
+              s.update(
+                index,
+                index + exp.length,
+                `new URL((import.meta.glob(${JSON.stringify(
+                  pattern,
+                )}, { eager: true, import: 'default', as: 'url' }))[${rawUrl}] || \`${builtUrl}\`, self.location)`,
```

Comment:
I think there's an extra comma after `pattern` and an extra pair of parentheses around `import.meta.glob` that we could remove to make the output a bit smaller. Also, I'm not sure if prettier is controlling the formatting here or not (I can't remember if it touches template strings), but if it's not then I'd tweak it a bit
```suggestion
                `new URL(
                   import.meta.glob(${JSON.stringify(pattern)},
                   { eager: true, import: 'default', as: 'url' }
                 )[${rawUrl}] || \`${builtUrl}\`, self.location)`,
```

Reply:
I have tested it and confirmed that Prettier does not format template strings. [See here](https://www.prettier.cn/playground/#N4Igxg9gdgLgprEAuEACABlOB3VBVAJQBkAKAHSlSoEsBbABwgCcYA6WuGAQ1YHMAbCACMSAEmAApAMoB5AHKsAzjCbUovagDMAniXpcY8JlACUAXwA0FKqmCo4XXnCZJUKgK5wLqOoxauAcgATOE0ud34YAO8uRUD3Jn4A1DNrEwBtcSYubDxEswBdVAAfYtQyMnRxIXdqSLz+VMrvRTh+TVZBMANqaBN0EAsQCHoYXqhFZFAuJiYIbAAFGYRJlC5+bC5tSaGhbLAAa04pLg4iNThkMP5W3f2jmCl9MDVeZA8vEDhaITggkKCRC46ncjjgADFmLQDGN1MgQOEYBBBiAABYwWj8ADqqOo8EUzzgUhWeOoADc8dp4WBFDsQGpWiwFtleNCrutbiAAFaKAAeUle-DgAEV3BB4Oybp99ExGfChFxfvwUfRVLAsdQgjBUcgABwABiGqogrSx2Xo8NVcEZZMuQwAjmL4MyRqsEYoALRYP5-FFMOCO6j+5mONlIa6c1q0ajvJieIaKQUip2XcMcz7cIQarU6pAAJiGKi4dVeAGEILQw19FABWFHuVoAFUVqwjnzJngAklAQrApGBVKMAII9qQwbRCyWtMxmIA).